### PR TITLE
TC Automation: SSE S3 TC Automation

### PR DESCRIPTION
**Pull request by sn-px**
**What this PR does / why we need it**:
https://portworx.atlassian.net/browse/PA-1742
Backup and restore from Backup location created with SSE-S3 and deny policy set from aws s3

**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PA-1742

**Special notes for your reviewer**:
Need to fix minor issue in clean up
Update Backup location looks like product issue , debugging is in progress will conclude tomorrow , so commented that step

Logs : https://aetos.pwx.purestorage.com/resultSet/testSetID/402862/testCaseID/2335349
Latest Logs : https://aetos.pwx.purestorage.com/resultSet/testSetID/405375/testCaseID/2350390
Jenkins logs :https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/new-byoc-pipeline/78/console

Latest log after SSE fix : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/new-byoc-pipeline/81/console

Also all test case run after adding validation flag : https://aetos.pwx.purestorage.com/resultSet/testSetID/407929

### CODE REVIEW

File: tests/backup/backup_sse_test.go [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+				log.InfoD("Remove deny policy from S3 bucket")
+				err = RemoveS3BucketPolicy(customBucketsWithDenyPolicy[1])
+				dash.VerifySafely(err, nil, fmt.Sprintf("Verify removal of deny policy from s3 bucket"))
+			})
+			Step("Taking backup of application after removal of deny policy and sse still set to true", func() {
+				log.InfoD("Taking backup of application after removal of deny policy and sse still set to true")
+				backupNameAfterRemovalOfDenyPolicy = fmt.Sprintf("%s-%s-%v", BackupNamePrefix, bkpNamespaces[0], time.Now().Unix())
+				appContextsToBackup = FilterAppContextsByNamespace(scheduledAppContexts, []string{bkpNamespaces[0]})
+				err = CreateBackupWithValidation(ctx, backupNameAfterRemovalOfDenyPolicy, SourceClusterName, newBackupLocationWithSseAfterRestart, backupLocationUID, appContextsToBackup, make(map[string]string), orgID, clusterUid, "", "", "", "")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creation and Validation of backup after removal of deny policy and sse still set to true [%s]", backupNameAfterRemovalOfDenyPolicy))
```
Comment (trenukarya-px) [inline tests/backup/backup_sse_test.go:370]: Is this expected to fail ? What is the expected error that we need to compare here in that case ?

File: tests/backup/backup_sse_test.go [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+					go func(scName string) {
+						defer GinkgoRecover()
+						defer wg.Done()
+						err = CreateRestore(restoreName, backupsWithSse[0], namespaceMap, SourceClusterName, orgID, ctx, storageClassMapping)
+						dash.VerifyFatal(err, nil, fmt.Sprintf("Restoring backup %v using storage class %v", backupName, scName))
+					}(scNames[i])
+				}
+				wg.Wait()
+			})
+			//Step("Update backup location backupLocationUidWithoutSse to sse true", func() {
```
Comment (trenukarya-px) [inline tests/backup/backup_sse_test.go:257]: why is this commented ?

File: tests/backup/backup_sse_test.go [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+					log.Infof("Updated S3 bucket policy - %s", globalAWSBucketName)
+				}
+			})
+			// Create BackupLocation with bucketWithPolicy and SSE set to false
+			Step("Create BackupLocation with bucketWithOutPolicy and SSE set to false", func() {
+				log.InfoD(fmt.Sprintf("Create BackupLocation with bucketWithOutPolicy and SSE set to false"))
+				cloudCredUidForBlWithoutSse = uuid.New()
+				cloudCredUidList = append(cloudCredUidList, cloudCredUID)
+				backupLocationUidWithoutSse = uuid.New()
+				credNameForBlWithoutSse = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
```
Comment (trenukarya-px) [inline tests/backup/backup_sse_test.go:134]: where are validating the backup failure with this backuplocation ?

File: tests/backup/backup_sse_test.go (hunk 2 above)
Comment (sn-px) [reply tests/backup/backup_sse_test.go:257]: It is due to product issue ,Once this update issue is fixed i will enable it
 https://portworx.atlassian.net/browse/PB-4687

File: tests/backup/backup_sse_test.go (hunk 3 above)
Comment (sn-px) [reply tests/backup/backup_sse_test.go:149]: Done

Comment (sn-px) [reply tests/backup/backup_sse_test.go:149]: Done ,Added negative scenario

File: deployments/deploy-ssh.sh [hunk 4]
```diff
@@ -721,6 +721,8 @@ spec:
       value: "${S3_POLICY_SID}"
     - name: S3_ENCRYPTION_POLICY
       value: "${S3_ENCRYPTION_POLICY}"
+    - name: PX_NAMESPACE
+      value: "${PX_NAMESPACE}"
```
Comment (mkoppal-px) [inline deployments/deploy-ssh.sh:725]: Why is this required?

Comment (sn-px) [reply deployments/deploy-ssh.sh:725]: If px is deployed in different namespace , restart stork will start failing .. so in case we have a setup while testing if px is deployed in different namespace we can pass the namespace and test or else we need to re-deploy the px

File: tests/common.go [hunk 5]
```diff
@@ -4353,6 +4364,83 @@ func CreateS3BackupLocationWithContext(name string, uid, cloudCred string, cloud
 	return nil
 }
 
+// CreateS3BackupLocationWithSseType creates backup location for S3 with SSE type
+func CreateS3BackupLocationWithSseType(name string, uid, cloudCred string, cloudCredUID string, bucketName string, orgID string, encryptionKey string, sseS3EncryptionType api.S3Config_Sse) error {
+	time.Sleep(60 * time.Second)
```
Comment (mkoppal-px) [inline tests/common.go:4369]: Please remove the sleep here

File: tests/common.go [hunk 6]
```diff
@@ -4353,6 +4364,83 @@ func CreateS3BackupLocationWithContext(name string, uid, cloudCred string, cloud
 	return nil
 }
 
+// CreateS3BackupLocationWithSseType creates backup location for S3 with SSE type
+func CreateS3BackupLocationWithSseType(name string, uid, cloudCred string, cloudCredUID string, bucketName string, orgID string, encryptionKey string, sseS3EncryptionType api.S3Config_Sse) error {
```
Comment (mkoppal-px) [inline tests/common.go:4368]: Why do we need a separate function for SSE Type. Can we not just add ` sseS3EncryptionType api.S3Config_Sse` as an optional argument in `CreateS3BackupLocation` itself?

File: tests/common.go [hunk 7]
```diff
@@ -4287,8 +4297,9 @@ func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID str
 			Uid:   uid,
 		},
 		BackupLocation: &api.BackupLocationInfo{
-			Path:          bucketName,
-			EncryptionKey: encryptionKey,
+			Path:                    bucketName,
+			EncryptionKey:           encryptionKey,
+			ValidateCloudCredential: validate,
```
Comment (mkoppal-px) [inline tests/common.go:4302]: Shouldn't you do the same in `CreateBackupLocationWithContext`

File: tests/backup/backup_sse_test.go [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+	. "github.com/portworx/torpedo/tests"
+	v1 "k8s.io/api/core/v1"
+	storageApi "k8s.io/api/storage/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ = Describe("{sseS3encryption}", func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:23]: Let the test case name be a bit more descriptive and follow the convention that is followed in other tests with the camel case

File: tests/backup/backup_sse_test.go [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+	)
+
+	storageClassMapping := make(map[string]string)
+	namespaceMap := make(map[string]string)
+	params := make(map[string]string)
+	k8sStorage := storage.Instance()
+	backupLocationMap := make(map[string]string)
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("sseS3encryption",
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:68]: Change this to use `StartPxBackupTorpedoTest`

File: tests/backup/backup_sse_test.go [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+			appContexts := ScheduleApplications(taskName)
+			for _, ctx := range appContexts {
+				ctx.ReadinessTimeout = appReadinessTimeout
+				namespace := GetAppNamespace(ctx, taskName)
+				bkpNamespaces = append(bkpNamespaces, namespace)
+				scheduledAppContexts = append(scheduledAppContexts, ctx)
+			}
+		}
+	})
+	It("sse-s3 encryption", func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:84]: The string should be meaningful in It block

File: tests/backup/backup_sse_test.go [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+	params := make(map[string]string)
+	k8sStorage := storage.Instance()
+	backupLocationMap := make(map[string]string)
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("sseS3encryption",
+			"Backup and restore from Backup location created with SSE-S3 and deny policy set from aws s3", nil, 93014)
+		log.InfoD("Deploy applications needed for backup")
+		scheduledAppContexts = make([]*scheduler.Context, 0)
+		numberOfNameSpace := 4
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:72]: Why do we need 4 namespaces?

Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:72]: Please follow the camel case correctly. A better variable name would be `namespaces` or `scaleFactor`.
`NameSpace` is not the right way. It should be `Namespace`.

File: tests/backup/backup_sse_test.go (hunk 2 above)
Comment (mkoppal-px) [reply tests/backup/backup_sse_test.go:273]: Please add a `TODO` with the bug ID

File: tests/backup/backup_sse_test.go [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+				log.InfoD(fmt.Sprintf("Creating source and destination cluster"))
+				err = CreateApplicationClusters(orgID, "", "", ctx)
+				dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+				clusterStatus, err = Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+				log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+				dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+				clusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+			})
+			// Create bucket without deny policy
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:105]: These comments are not needed. The step itself is self-explanatory

File: tests/backup/backup_sse_test.go [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+				customBuckets = append(customBuckets, customBucketWithOutPolicy)
+			})
+			// Create bucket with deny policy
+			Step("Create bucket with deny policy", func() {
+				log.InfoD(fmt.Sprintf("Create bucket with deny policy"))
+				numberOfBucketsWithDenyPolicy := 2
+				bucketWithDenyPolicy := "sse-bucket-with-deny-policy"
+				for i := 0; i < numberOfBucketsWithDenyPolicy; i++ {
+					customBucketWithDenyPolicy := GetCustomBucketName(provider, bucketWithDenyPolicy)
+					policy, err := GenerateS3BucketPolicy("DenyNonAES256Uploads", "s3:x-amz-server-side-encryption=AES256", customBucketWithDenyPolicy)
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:120]: Please use consts here instead of directly adding the string

File: tests/backup/backup_sse_test.go [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+					log.FailOnError(err, "Failed to generate s3 bucket policy check for the correctness of policy parameters")
+					err = UpdateS3BucketPolicy(customBucketWithDenyPolicy, policy)
+					log.FailOnError(err, "Failed to apply bucket policy")
+					customBucketsWithDenyPolicy = append(customBucketsWithDenyPolicy, customBucketWithDenyPolicy)
+					customBuckets = append(customBuckets, customBucketWithDenyPolicy)
+					time.Sleep(5 * time.Second)
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:126]: Remove all the time.Sleep

File: tests/backup/backup_sse_test.go [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+				cloudCredUidForBlWithoutSse = uuid.New()
+				cloudCredUidList = append(cloudCredUidList, cloudCredUID)
+				backupLocationUidWithoutSse = uuid.New()
+				credNameForBlWithoutSse = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				err := CreateCloudCredential(provider, credNameForBlWithoutSse, cloudCredUidForBlWithoutSse, orgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of cloud credential named [%s] for org [%s] with [%s] as provider", credName, orgID, provider))
+				log.InfoD("Created Cloud Credentials with name - %s", credNameForBlWithoutSse)
+				backupLocationWithoutSse = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				err = CreateBackupLocation(provider, backupLocationWithoutSse, backupLocationUidWithoutSse, credNameForBlWithoutSse, cloudCredUidForBlWithoutSse, customBucketsWithDenyPolicy[0], orgID, "", true)
+				dash.VerifyFatal(strings.Contains(err.Error(), "AccessDenied"), true, fmt.Sprintf("Verifing backup location creation fail in case of bucket with deny policy and SSE set to false"))
```
Comment (mkoppal-px) [inline tests/backup/backup_sse_test.go:142]: Fix the typo `Verifing` and also print the error message

File: tests/common.go (hunk 6 above)
Comment (sn-px) [reply tests/common.go:4368]: In that case all test needs to fetch sseS3EncryptionType and pass to CreateS3BackupLocation in case of global run

File: tests/common.go [hunk 16]
```diff
@@ -4022,11 +4022,11 @@ func CreateApplicationClusters(orgID string, cloudName string, uid string, ctx c
 }
 
 // CreateBackupLocation creates backup location
-func CreateBackupLocation(provider, name, uid, credName, credUID, bucketName, orgID string, encryptionKey string) error {
+func CreateBackupLocation(provider, name, uid, credName, credUID, bucketName, orgID, encryptionKey string, validate bool) error {
 	var err error
 	switch provider {
 	case drivers.ProviderAws:
-		err = CreateS3BackupLocation(name, uid, credName, credUID, bucketName, orgID, encryptionKey)
+		err = CreateS3BackupLocation(name, uid, credName, credUID, bucketName, orgID, encryptionKey, validate)
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/common.go:4029]: yes for nfs and also we nned to pass for other providers as well

File: deployments/deploy-ssh.sh (hunk 4 above)
Comment (mkoppal-px) [reply deployments/deploy-ssh.sh:725]: There is already a function called `getExecPodNamespace` which gives you the namespace where portworx is running

File: tests/common.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (mkoppal-px) [reply tests/common.go:4029]: Please make the change

File: tests/common.go (hunk 6 above)
Comment (mkoppal-px) [reply tests/common.go:4368]: No. The `sseS3EncryptionType` will be optional. Only SSE specific tests will provide it. For others the function signature will not change

File: tests/backup/backup_sse_test.go (hunk 11 above)
Comment (mkoppal-px) [reply tests/backup/backup_sse_test.go:72]: Ok

File: tests/common.go (hunk 6 above)
Comment (sn-px) [reply tests/common.go:4368]: yes correct in case of specific case and ideally we should follow that, but in this case of global run we are reading encryption type from the env variable , tests need to read from the env and pass to createS3Backup location 
we can discuss on this

Comment (trenukarya-px) [reply tests/common.go:4368]: We can read the argument or env variable both in function right ? If argument not passed, we can check env variable.

File: deployments/deploy-ssh.sh (hunk 4 above)
Comment (sn-px) [reply deployments/deploy-ssh.sh:725]: yes it makes sense , currently 3 tests uses DeletePodWithLabelInNamespace(getPXNamespace(), storkLabel)
Shall i go ahead and modify all?

File: tests/backup/backup_sse_test.go [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+				bkpNamespacesForWithoutSse := bkpNamespaces[:midpoint]
+				for _, backupNameSpace := range bkpNamespacesForWithoutSse {
+					backupName = fmt.Sprintf("%s-%s-%s-%v", BackupNamePrefix, backupNameSpace, RandomString(randomStringLength), time.Now().Unix())
+					wg.Add(1)
+					go func(backupNameSpace string, backupName string) {
+						defer GinkgoRecover()
+						defer wg.Done()
+						appContextsToBackup = FilterAppContextsByNamespace(scheduledAppContexts, []string{backupNameSpace})
+						err = CreateBackupWithValidation(ctx, backupName, SourceClusterName, backupLocationWithoutSse, backupLocationUidWithoutSse, appContextsToBackup, make(map[string]string), orgID, clusterUid, "", "", "", "")
+						dash.VerifyFatal(err, nil, fmt.Sprintf("Creation and Validation of backup [%s]", backupName))
```
Comment (trenukarya-px) [inline tests/backup/backup_sse_test.go:184]: Error is nil here since it is "backupLocationUidWithoutSse' ?

File: tests/backup/backup_sse_test.go [hunk 18]
```diff
@@ -0,0 +1,425 @@
...
+					}(backupNameSpace, backupName)
+					backupsWithSse = append(backupsWithSse, backupName)
+				}
+				wg.Wait()
+			})
+			Step("Create restore with replace policy set to retain", func() {
+				log.InfoD("Create restore with replace policy set to retain")
+				restoreName := fmt.Sprintf("restore-with-replace-%s-%s-%v", RestoreNamePrefix, RandomString(randomStringLength), time.Now().Unix())
+				err = CreateRestoreWithReplacePolicy(restoreName, backupsWithSse[0], make(map[string]string), SourceClusterName, orgID, ctx, make(map[string]string), 2)
```
Comment (kphalgun-px) [inline tests/backup/backup_sse_test.go:212]: Validation?

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/common.go [hunk 19]
```diff
@@ -4418,14 +4436,50 @@ func CreateS3BackupLocationWithContext(name string, uid, cloudCred string, cloud
 	}
 
 	_, err = backupDriver.CreateBackupLocation(ctx, bLocationCreateReq)
+	if err != nil {
+		return fmt.Errorf("failed to create backup location: %v", err)
+	}
+	return nil
+}
+
+// UpdateS3BackupLocation with the provided values
```
Comment (kshithijiyer-px) [inline tests/common.go:4445]: // UpdateS3BackupLocation updates s3 backup location with the provided values

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

File: tests/common.go [hunk 20]
```diff
@@ -4338,18 +4349,23 @@ func CreateCloudCredential(provider, credName string, uid, orgID string, ctx con
 }
 
 // CreateS3BackupLocation creates backup location for S3
-func CreateS3BackupLocation(name string, uid, cloudCred string, cloudCredUID string, bucketName string, orgID string, encryptionKey string) error {
+func CreateS3BackupLocation(name, uid, cloudCred, cloudCredUID, bucketName, orgID, encryptionKey string, validate bool, sseS3EncryptionType ...api.S3Config_Sse) error {
```
Comment (mkoppal-px) [inline tests/common.go:4352]: The optional parameter needs to be added to the function `CreateS3BackupLocationWithContext`

File: tests/common.go [hunk 21]
```diff
@@ -4338,18 +4349,23 @@ func CreateCloudCredential(provider, credName string, uid, orgID string, ctx con
...
 	_, _, endpoint, region, disableSSLBool := s3utils.GetAWSDetailsFromEnv()
-	// Get SSE S3 Encryption Type
-	sseS3EncryptionType, err := GetSseS3EncryptionType()
-	if err != nil {
-		return err
+	// Initialize a new variable to hold the SSE S3 Encryption Type
+	var sseType api.S3Config_Sse
+	if len(sseS3EncryptionType) == 0 {
+		// If sseS3EncryptionType Type parameter is not passed , then take it from environment variable
+		sseType, err = GetSseS3EncryptionType()
```
Comment (mkoppal-px) [inline tests/common.go:4363]: Will this not throw error if env variable `S3_SSE_TYPE` is not set?

**Review by mkoppal-px — APPROVED** [review]
LGTM
